src/trajectory/util.cpp

```diff
@@ -0,0 +1,318 @@
+      stateSpace, _inputTrajectory.getStartTime());
+
+  Eigen::VectorXd currentVec, nextVec;
+  for (std::size_t iwaypoint = 0; iwaypoint < numWaypoints - 1; ++iwaypoint)
+  {
+    const auto currentState = _inputTrajectory.getWaypoint(iwaypoint);
+    double currentTime = _inputTrajectory.getWaypointTime(iwaypoint);
+    const auto nextState = _inputTrajectory.getWaypoint(iwaypoint + 1);
+    double nextTime = _inputTrajectory.getWaypointTime(iwaypoint + 1);
+
+    stateSpace->logMap(currentState, currentVec);
+    stateSpace->logMap(nextState, nextVec);
+
+    // Compute the spline coefficients for this segment of the trajectory.
+    LinearSplineProblem problem(
+        Vector2d(0, nextTime - currentTime), 2, dimension);
+    problem.addConstantConstraint(0, 0, Eigen::VectorXd::Zero(dimension));
+    problem.addConstantConstraint(1, 0, nextVec - currentVec);
+    const auto spline = problem.fit();
+
+    assert(spline.getCoefficients().size() == 1);
+    const auto& coefficients = spline.getCoefficients().front();
+    outputTrajectory->addSegment(
+        coefficients, nextTime - currentTime, currentState);
+  }
+
+  return outputTrajectory;
+}
+
+//==============================================================================
+double findTimeOfClosetStateOnTrajectory(
+    const aikido::trajectory::Trajectory* traj,
+    const Eigen::VectorXd& referenceState,
+    double timeStep)
+{
+  if (traj == nullptr)
+    throw std::runtime_error("Traj is nullptr");
+  auto stateSpace = traj->getStateSpace();
+  std::size_t configSize = referenceState.size();
+  if (configSize != stateSpace->getDimension())
+    throw std::runtime_error("Dimension mismatch");
+
+  double findTime = traj->getStartTime();
+  double minDist = std::numeric_limits<double>::max();
+
+  aikido::common::StepSequence sequence(
+      timeStep, true, true, traj->getStartTime(), traj->getEndTime());
+
+  auto currState = stateSpace->createState();
+  Eigen::VectorXd currPos(stateSpace->getDimension());
+  for (std::size_t i = 0; i < sequence.getLength(); i++)
+  {
+    double currTime = sequence[i];
+    traj->evaluate(currTime, currState);
+    stateSpace->logMap(currState, currPos);
+
+    double currDist = (referenceState - currPos).norm();
+    if (currDist < minDist)
+    {
+      findTime = currTime;
+      minDist = currDist;
+    }
+  }
+  return findTime;
+}
+
+//==============================================================================
+std::unique_ptr<aikido::trajectory::Spline> createPartialTrajectory(
+    const aikido::trajectory::Spline& traj, double partialStartTime)
+{
+  if (partialStartTime < traj.getStartTime()
+      || partialStartTime > traj.getEndTime())
+    throw std::runtime_error("Wrong partial start time");
+
+  auto stateSpace = traj.getStateSpace();
+  const std::size_t dimension = stateSpace->getDimension();
+  auto outputTrajectory = make_unique<aikido::trajectory::Spline>(
+      stateSpace, traj.getStartTime());
```

Comment:
Why does it use `traj.getStartTime()` instead of `partialStartTime`?

Reply:
It is like cutting a partial trajectory from a given start to the end; then shifting it to the same start time.